src/lamp_py/performance_manager/alerts.py

```diff
@@ -0,0 +1,478 @@
+        if self.local_exists:
+            existing_alerts = pq.read_table(
+                self.local_path, columns=columns
+            ).to_pandas()
+            existing_alerts = existing_alerts.drop_duplicates()
+            return existing_alerts
+
+        return pandas.DataFrame(columns=columns)
+
+    def append_new_records(self, alerts: pandas.DataFrame) -> None:
+        """
+        append alerts to the end of the local parquet file using batches to keep memory usage lower
+        """
+        self.new_data = True
+        alerts = alerts.reset_index(drop=True)
+        alerts_table = pyarrow.Table.from_pandas(
+            alerts, schema=self.parquet_schema
+        )
+
+        # if there is no local file, write the alerts to the local file and exit
+        if not self.local_exists:
+            with pq.ParquetWriter(
+                self.local_path, schema=self.parquet_schema
+            ) as writer:
+                writer.write_table(alerts_table)
+
+            self.local_exists = True
+            return
+
+        # write new alerts to a new file and then join them with the local file in batches
+        new_path = os.path.join("/tmp", "new_alerts.parquet")
+        joined_path = os.path.join("/tmp", "joined_alerts.parquet")
+
+        with pq.ParquetWriter(new_path, schema=self.parquet_schema) as writer:
+            writer.write_table(alerts_table)
+
+        joined_batches = pd.dataset(
+            [
+                pd.dataset(self.local_path),
+                pd.dataset(new_path),
+            ]
+        ).to_batches()
+
+        with pq.ParquetWriter(
+            joined_path, schema=self.parquet_schema
+        ) as writer:
+            for batch in joined_batches:
+                writer.write_batch(batch)
+
+        # the joined file becomes the local file and we can delete the new and
+        # joined files
+        os.replace(joined_path, self.local_path)
+        os.remove(new_path)
+
+    def download_data(self) -> bool:
+        """
+        download the s3 parquet file to a temp location if the versions match.
+        return True if the file was downloaded successfully, False otherwise.
+        """
+        try:
+            lamp_version = object_metadata(self.s3_path).get(
+                self.version_key, ""
+            )
+
+            if lamp_version != self.file_version:
+                return False
+
+            return download_file(
+                object_path=self.s3_path,
+                file_name=self.local_path,
+            )
+
+        except Exception:
+            return False
+
+    def upload_data(self) -> None:
+        """
+        upload the local parquet file to s3 if new data was added
+        """
+        if self.new_data:
+            upload_file(
+                file_name=self.local_path,
+                object_path=self.s3_path,
+                extra_args={"Metadata": {self.version_key: self.file_version}},
+            )
+
+
+def extract_alerts(
```

Comment:
the flow for the script is
* download existing alerts
* get the unprocessed alert files
* read them, pulling out the columns we care for and drop duplicates and alerts we have already recorded
* transform translations
* transform timestamps
* explode around informed entity
* append to existing alerts and upload

downloading, appending, and uploading is handled by the previous class.